bot.go

```diff
@@ -86,54 +95,99 @@ func (b *Bot) Join(s *discordgo.Session, m *discordgo.MessageCreate, name string
 	if c, ok := b.channels[m.ChannelID]; ok {
 		if mod, ok := c.Mods[name]; ok {
 			g := GameIdentifier{m.ChannelID, name}
-			if _, ok := bot.games[g]; !ok {
-				bot.games[g] = Game{Players: make(map[string]bool), Blue: make(map[string]bool), Red: make(map[string]bool), mutex: new(sync.Mutex)}
+			if _, ok := b.games[g]; !ok {
+				b.games[g] = Game{Players: make(map[Player]bool), Blue: make(map[string]bool), Red: make(map[string]bool), mutex: new(sync.Mutex)}
 			}
-			bot.games[g].mutex.Lock()
-			defer bot.games[g].mutex.Unlock()
-			if len(bot.games[g].Players) == mod.MaxPlayers {
+			b.games[g].mutex.Lock()
+			defer b.games[g].mutex.Unlock()
+			
+			if len(b.games[g].Players) == mod.MaxPlayers {
 				return
 			}
-			bot.games[g].Players[m.Author.Username] = true
+			if b.games[g].PlayerWithName(m.Author.Username) != nil {
+				return
+			}
+			player := Player { m.Author.Username, false }
+			b.games[g].Players[player] = true
 			// TODO: Add logic to start a game.
 		}
 	}
 }
 
+func (b *Bot) J(s *discordgo.Session, m *discordgo.MessageCreate, name string) {
+	b.Join(s, m, name)
+}
+
+func (b *Bot) Jp(s *discordgo.Session, m *discordgo.MessageCreate, name string) {
+	b.Joinpm(s, m, name)
+}
+
+func (b *Bot) Joinpm(s *discordgo.Session, m *discordgo.MessageCreate, name string) {
+	b.Join(s, m, name)
+	b.Pm(s, m, name)
+}
+
+
+func (b *Bot) Pm(s *discordgo.Session, m *discordgo.MessageCreate, name string) {
+	if c, ok := b.channels[m.ChannelID]; ok {
+		if _, ok := c.Mods[name]; ok {
+			g := GameIdentifier{m.ChannelID, name}
+			if _, ok := b.games[g]; !ok {
+				return
+			}
+
+			for player := range b.games[g].Players {
```

Comment:
No need to do O(n) lookup, it's a dict

Reply:
@patzar currently doing this would work to only update a Player if they don't already have PM set:

```
player := Player { m.Author.Username, false }
if _, ok := b.games[g].Players[player]; ok {
	delete(b.games[g].Players, player)
	newPlayer := Player { m.Author.Username, true }
	b.games[g].Players[newPlayer] = true				
	s.MessageReactionAdd(m.ChannelID, m.Message.ID, "✅")
}
```

However once we add more fields to Player that will stop working. For example adding a `.tag nomic` feature would require a change to `Player` to store that person's tag as nomic. Personally I don't much care about tagging but it's an existing feature that people use quite often.

I don't believe Go allows for defining custom hash values or equality functions on structs so we kinda need to O(n) these things.

I am going to update this to use `b.games[g].PlayerWithName(m.Author.Username)` unless you have some strong objections or can point me at a cleaner way to do this

Reply:
maybe the player map can be keyed on their username and hold PlayerMetadata struct with NotifyOnPM/Tag/etc?